caput/mpiarray.py

```diff
@@ -1602,98 +1609,203 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+                    ) from e
+        else:
+            out_args = [None] * ufunc.nout
 
-        if "reduce" in method and (
-            results[0].shape or getattr(outputs[0], "shape", None)
-        ):
-            # reduction methods eliminate axes, so the distributed axis
-            # might need to be recalculated
-            # except when the user explicitly specifies keepdims
-            if not kwargs.get("keepdims", False) and (kwargs["axis"] < dist_axis):
-                dist_axis -= 1
+        results = _ensure_list(
+            super().__array_ufunc__(ufunc, method, *_mpi_to_ndarray(inputs), **kwargs)
+        )
 
+        # Convert any outputs back into valid MPIArrays if required
         ret = []
+        for res, out in zip(results, out_args):
+            # If out is None a new *numpy* array was initialised for the results. We
+            # need to turn this into an MPIArray using the inferrred distribution
+            if out is None:
+                # Result is a scalar, so we want to return this as a 1D distributed
+                # vector
+                if not res.shape:
+                    res = res.reshape(1)
+
+                # Convert into an an MPIArray
+                out = MPIArray._view_from_data_and_params(
+                    res, new_dist_axis, global_length, offset, comm
+                )
 
-        for result, output in zip(results, outputs):
-            # case: results were placed in the array specified by `out`; return as is
-            if output is not None:
-                if hasattr(output, "axis") and output.axis != dist_axis:
-                    raise AxisException(
-                        "provided output MPIArray's distributed axis is not consistent "
-                        f"with expected output distributed axis. Expected {dist_axis}; "
-                        f"Actual {output.axis}"
-                    )
-                ret.append(output)
-            else:
-                # case: the result is an ndarray; wrap it into an MPIArray
-                if result.shape:
-                    ret.append(MPIArray.wrap(result, axis=dist_axis))
-                # case: result is a scalar; convert to 1-d vector, distributed across
-                # axis 0
-                else:
-                    ret.append(MPIArray.wrap(np.reshape(result, (1,)), axis=0))
+            ret.append(out)
 
         return ret[0] if len(ret) == 1 else tuple(ret)
 
+    def _array_ufunc_call(self, ufunc, *inputs, **kwargs):
+        # Validate and infer the final distributed axis for a standard ufunc call
+
+        # For a 'call' the highest dimensional array should determine the dimensionality
+        # of the output, so we find that
+        max_dim_input = inputs[
+            np.argmax(
+                [inp.ndim if isinstance(inp, np.ndarray) else -1 for inp in inputs]
+            )
+        ]
+
+        if not isinstance(max_dim_input, MPIArray):
```

Comment:
This appear to cause issues in unintended places. Tested in test_daily pipeline (python 3.10.2, numpy 1.22.2) results in the following error:
```
   139.7s [MPI 0/4] - DEBUG    caput.pipeline: Task ComputeSystemSensitivity calling 'next()'.
   139.7s [MPI 0/4] - INFO     draco.analysis.sensitivity.ComputeSystemSensitivity: Starting next for task ComputeSystemSensitivity
...
  File "/project/6003614/lgray/chime_pipeline_dev/code/draco/draco/analysis/sensitivity.py", line 213, in process
    radiometer[:, pp, :] += nsq * auto[:, ii, :] * auto[:, jj, :]
  File "/project/6003614/lgray/chime_pipeline_dev/code/caput/caput/mpiarray.py", line 1686, in __array_ufunc__
    new_dist_axis, global_length, offset = _validation_fn(ufunc, *inputs, **kwargs)
  File "/project/6003614/lgray/chime_pipeline_dev/code/caput/caput/mpiarray.py", line 1741, in _array_ufunc_call
    raise ValueError("Highest dimensional input must be an MPIArray.")
ValueError: Highest dimensional input must be an MPIArray.
```

Reply:
Thanks! I hadn't yet gotten around to running `test_daily` on it yet. I'll fix that issue and try it again myself.